### DISCUSSION TITLE
Using actions to spin the robot (ROS2)

### DISCUSSION THREAD

**Discussion by Dee4Sniper** [iRobotEducation/create3_docs#459] [Questions] [answered]
### How are you connecting to your Create 3?

USB Ethernet (ROS 2)

### Computer(s) Model(s) and Operating System(s)

NavQ+ on Ubuntu 22

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.3

### Which RMW is your robot running?

CycloneDDS

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

I have been trying to get the robot to move, but I have been unsuccessful... I tried using the python web playground that works. I am trying to use the NavQ+ to send an action to the robot for it to go straight or turn. None of the examples work:

`ros2 action send_goal /rotate_angle irobot_create_msgs/action/RotateAngle "{angle: 1.57,max_rotation_speed: 0.5}"`

I tried creating my own action file then running it but it also does not do anything... It doesnt seem to send anything to the robot. I can read all the topics fine using `ros2 topic list` even echoing them, but when I try anything that publishes to the robot it doesnt work.

What am I doing wrong? Is there something I am missing?

**Comment by alsora** [comment]
Please clarify what you mean by "it doesn't work".
Is the action goal being accepted? Is the robot not moving?

Also please include the output of the  `ros2 action...` command and a console log from the robot

**Reply by Dee4Sniper** [reply]
The robot does not seem to accept the action goals. I use the line:

`ros2 action send_goal /drive_distance irobot_create_msgs/action/DriveDistance "{distance: 0.5,max_translation_speed: 0.15}"`

It outputs:

`Waiting for an action server to become available...
Sending goal:
     distance: 0.5
max_translation_speed: 0.15
`
This just gets stuck here and the robot doesnt do anything... 

Below is part of the console logs (full log is attached):

`Nov 21 15:00:07 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 syslog.info syslogd started: BusyBox v1.30.1
Nov 21 15:00:07 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 kern.info kernel: [    0.111119] Bluetooth: L2CAP socket layer initialized
Nov 21 15:00:07 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 kern.info kernel: [    0.111152] Bluetooth: SCO socket layer initialized
Nov 21 15:00:07 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 kern.info kernel: [    0.113454] clocksource: Switched to clocksource MIPS
Nov 21 15:00:07 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 kern.info kernel: [    0.118434] NET: Registered protocol family 2
`

[messages.txt](https://github.com/iRobotEducation/create3_docs/files/13448562/messages.txt)

**Reply by Dee4Sniper** [reply]
I have got these errors:

![image](https://github.com/iRobotEducation/create3_docs/assets/59607775/78cef307-5a91-493f-9346-054cc9964eb6)

**Reply by shamlian** [reply]
Those errors are related to the robot not having an active Wi-Fi connection and should not cause a problem directly (especially if the robot is intentionally disconnected). See below for how this can affect Cyclone 0.9.x.

**Comment by shamlian** [comment]
I don't know, but I suspect that you should double check that the middleware being used on the robot and the middleware used on your computer are the same; a mismatch often causes this (topics list but actions don't act).
On your computer: echo $RMW_IMPLEMENATION
On your robot: check application settings [(docs here)](https://iroboteducation.github.io/create3_docs/webserver/application/), or provide the robot log (http://robot_ip/logs_raw)

**Reply by Dee4Sniper** [reply]
Both the computer and the robot are running on CycloneDDS so i am guessing the issue isnt with the middleware.

On computer I set this on my `.bashrc` file:

![image](https://github.com/iRobotEducation/create3_docs/assets/59607775/dd18fbcb-267c-4a15-8945-13ebf6264b3c)

On the application settings of the robot:

![image](https://github.com/iRobotEducation/create3_docs/assets/59607775/0b9fa04c-c99b-439c-b6dd-055a15559176)


I do get this as the logs:

`Nov 23 10:15:34 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 daemon.warn network[1469]: [WARN] (src/gateway_pinger.c:132) no address count, 12
Nov 23 10:15:36 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 user.notice create-platform: [INFO] [1700734536.837551456] [system_health]: CPU usage: max 61 [%] mean 56 [%] RAM usage: 32/59 [MB]
Nov 23 10:15:38 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 user.notice create-platform: 1700734538.656922 [0]        tev: ddsi_udp_conn_write to udp/239.255.0.1:7400 failed with retcode -3
Nov 23 10:15:38 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 user.notice create-platform: 1700734538.658071 [0]        tev: ddsi_udp_conn_write to udp/10.57.47.26:7412 failed with retcode -3
Nov 23 10:15:46 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 user.notice create-platform: 1700734546.657515 [0]        tev: ddsi_udp_conn_write to udp/239.255.0.1:7400 failed with retcode -3
Nov 23 10:15:46 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 user.notice create-platform: 1700734546.658203 [0]        tev: ddsi_udp_conn_write to udp/10.57.47.26:7412 failed with retcode -3
Nov 23 10:15:54 iRobot-9F2C6F113AFE4C6D85C16253FC7CB150 daemon.notice network[1469]: [INFO] (src/gateway_pinger.c:121) wifi is disconnected`

[messages (1).txt](https://github.com/iRobotEducation/create3_docs/files/13448662/messages.1.txt)

**Reply by shamlian** [reply]
The `tev: ddsi_udp_conn_write to udp/10.57.47.26:7412 failed with retcode -3 ` is Cyclone trying to use the Wi-Fi interface, but it appears that interface is not connected. @bperseghetti posted a workaround below that I think will help.

**Comment by bperseghetti** [comment, marked as answer]
Interesting, looks like it might be requiring the wifi network as one of the network interface devices but failing to use it.

Right now the NavQPlus uses the current [humble version of CycloneDDS (0.10.x)](https://github.com/ros2/ros2/pull/1404) which has a slightly different config YAML for allowing "presence required". The current version of CycloneDDS on the Create3 is 0.9.x. I'm not aware of any reason why they should not be able to interoperate, but having them on the same version would be more ideal.

One thing to consider is also setting the [CycloneDDS config on the Create3](https://iroboteducation.github.io/create3_docs/webserver/rmw-profile-override/#example-for-cyclone-dds) to NOT use wifi

IE:
```
<CycloneDDS>
   <Domain>
     <General>
        <NetworkInterfaceAddress>usb0</NetworkInterfaceAddress>
    </General>
   </Domain>
</CycloneDDS>
```

**Reply by shamlian** [reply]
I agree, it seems like the logic to select the correct configuration file for Cyclone is choosing to include the Wi-Fi interface, and when the Wi-Fi interface drops out (or doesn't come up), Cyclone is throwing a bunch of errors. In the case that you care to only use the `usb0` interface, overriding the configuration XML with this one should do the trick.

**Reply by Dee4Sniper** [reply]
Thanks for the help it is working now!
